### PR TITLE
feat(theme) theme development

### PR DESCRIPTION
**Pull request by rbioteau**
Closes BST-211

### CODE REVIEW

File: md/customize-living-application-theme.md [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+
+This page explains how to create a theme project in the Studio.
+
+A theme is an archive of files used to define and share CSS definitions and images for an application.
+It enables you to specify the same style for all pages and layout of an application.
+Each application can have its own theme.
+
+## Prerequisites
+
+* An Internet connection to fetch maven and npm repositories
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]: Maven

File: md/customize-living-application-theme.md [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+* Maven basics
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
* Maven basic knowledge
```

File: md/customize-living-application-theme.md [hunk 3]
```diff
@@ -0,0 +1,327 @@
...
+* CSS basics
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
* CSS basic knowledge
```

File: md/customize-living-application-theme.md [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+
+## Create a Theme project
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
## Create a theme project
```

File: md/customize-living-application-theme.md [hunk 5]
```diff
@@ -0,0 +1,327 @@
...
+
+When you create a new theme (Right click in Project Explorer > New > Theme...), you start with a default project layout based on [Sass](https://sass-lang.com/) preprocessor and SCSS syntax.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
When you create a new theme (right click in Project Explorer > New > Theme...), you start with a default project theme based on [Sass](https://sass-lang.com/) preprocessor and SCSS syntax.
```

File: md/customize-living-application-theme.md [hunk 6]
```diff
@@ -0,0 +1,327 @@
...
+
+::: info
+**Note:** Sass is a css preprocessor and SCSS syntax is a superset of CSS 3. You can just write classic css 3 in a .scss file.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
**Note:** Sass is a CSS preprocessor and SCSS syntax is a superset of CSS 3. You can just write classic CSS 3 in a .scss file.
```

File: md/customize-living-application-theme.md [hunk 7]
```diff
@@ -0,0 +1,327 @@
...
+:::
+
+### Default project layout
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
### Default project theme
```

File: md/customize-living-application-theme.md [hunk 8]
```diff
@@ -0,0 +1,327 @@
...
+* An Internet connection to fetch Maven and NPM repositories
+* Maven basic knowledge
+* CSS basic knowledge
+
+## Create a theme project
+
+When you create a new theme (right click in Project Explorer > New > Theme...), you start with a default project theme based on [Sass](https://sass-lang.com/) preprocessor and SCSS syntax.
+
+::: info
+**Note:** Sass is a CSS preprocessor and SCSS syntax is a superset of CSS 3. You can just write classic CSS 3 in a .scss file.
```
Comment (abirembaut) [inline md/customize-living-application-theme.md]: common usage is to write CSS3 without whitespace

File: md/customize-living-application-theme.md [hunk 9]
```diff
@@ -0,0 +1,327 @@
...
+:::
+
+### Default project theme
```
Comment (abirembaut) [inline md/customize-living-application-theme.md]: I would rather say "Default theme project"

File: md/customize-living-application-theme.md [hunk 10]
```diff
@@ -0,0 +1,327 @@
...
+|	| index.html //A static test page to preview your theme
+| content.xml //Maven assembly descriptor
+| package-lock.json //NPM dependencies version lock, ensure a consistent install and compatible dependencies
+| package.json //NPM descriptor
+| page.properties //Custom page descriptor
+| pom.xml //Maven project descriptor
+```
+
+This project is using [Sass](https://sass-lang.com/) to easily create and customize a bootstrap *3.3.7* theme.
+To compile the scss it relies on [Node](https://nodejs.org/en/) and [NPM](https://www.npmjs.com/). See the package.json file below for more detailed.
```
Comment (abirembaut) [inline md/customize-living-application-theme.md]: more detail**s**

File: md/customize-living-application-theme.md [hunk 11]
```diff
@@ -0,0 +1,327 @@
...
+    "autoprefixer": "9.5.0",
+    "bootstrap-sass": "3.3.7" //Supported version of bootstrap in Bonita
+  }
+}
+```
+
+By default, a _build_ npm script is defined. It runs `node-sass` to compile the `src/scss/main.scss` file. The build command includes the bootstrap-sass stylesheets in order to have clean `@import` statements in the _*.scss_ files.
+In addition, *postcss-cli* and *autoprefixer* are used to add vendor prefixes for a better browser compatibility.
+
+The maven descriptor is responsible to run the npm build and package the result as a Theme custom page archive. See the pom.xml file below for more detailed.
```
Comment (abirembaut) [inline md/customize-living-application-theme.md]: responsible for running the npm
more details

File: md/customize-living-application-theme.md [hunk 12]
```diff
@@ -0,0 +1,327 @@
...
+						</configuration>
+					</execution>
+				</executions>
+			</plugin>
+		</plugins>
+	</build>
+</project>
+```
+
+The `artifactId`, `name` and `description` are used to defined the theme metadata (the name used in the URL, the display name, and a description) in the `page.properties`. NodeJS and NPM version are fixed in the `properties` section. The `frontend-maven-plugin` will locally install and use these versions even if you have  NodeJS and NPM already installed in your environment. See the [plugin github repository](https://github.com/eirslett/frontend-maven-plugin) for more informations.
```
Comment (abirembaut) [inline md/customize-living-application-theme.md]: used to defin**e**

File: md/customize-living-application-theme.md [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+
+$panel-success-heading-bg: $brand-success;
+$panel-info-heading-bg: $brand-info;
+$panel-warning-heading-bg: $brand-warning;
+$panel-danger-heading-bg: $brand-danger;
+
+//Glyphicons fonts
+$icon-font-path: "./fonts/"; // path relative to the theme.css file in the dist folder
+```
+All the variables defined in this file are used by *bootstrap-sass*. You can look the following scss file `node_modules/bootstratp-sass/assets/stylesheets/bootstrap/_variable.scss` to discover all available variables.
```
Comment (abirembaut) [inline md/customize-living-application-theme.md]: you can look **at** the following

**Review by abirembaut — COMMENTED** [review]
:+1: Just a few comments

File: md/customize-living-application-theme.md [hunk 14]
```diff
@@ -0,0 +1,327 @@
...
+|	| README.md
+|	| index.html //A static test page to preview your theme
+| content.xml //Maven assembly descriptor
+| package-lock.json //NPM dependencies version lock, ensure a consistent install and compatible dependencies
+| package.json //NPM descriptor
+| page.properties //Custom page descriptor
+| pom.xml //Maven project descriptor
+```
+
+This project is using [Sass](https://sass-lang.com/) to easily create and customize a bootstrap *3.3.7* theme.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
This project is using [Sass](https://sass-lang.com/) to easily create and customize a Bootstrap *3.3.7* theme.
```

File: md/customize-living-application-theme.md [hunk 15]
```diff
@@ -0,0 +1,327 @@
...
+To compile the scss it relies on [Node](https://nodejs.org/en/) and [NPM](https://www.npmjs.com/). See the package.json file below for more detailed.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
To compile the SCSS, it relies on [Node](https://nodejs.org/en/) and [npm](https://www.npmjs.com/). See the package.json file below for more details.
```

File: md/customize-living-application-theme.md [hunk 16]
```diff
@@ -0,0 +1,327 @@
...
+    "autoprefixer": "9.5.0",
+    "bootstrap-sass": "3.3.7" //Supported version of bootstrap in Bonita
+  }
+}
+```
+
+By default, a _build_ npm script is defined. It runs `node-sass` to compile the `src/scss/main.scss` file. The build command includes the bootstrap-sass stylesheets in order to have clean `@import` statements in the _*.scss_ files.
+In addition, *postcss-cli* and *autoprefixer* are used to add vendor prefixes for a better browser compatibility.
+
+The maven descriptor is responsible to run the npm build and package the result as a Theme custom page archive. See the pom.xml file below for more detailed.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
The Maven descriptor is responsible for running the npm build and package the result as a Theme custom page archive. See the pom.xml file below for more details.  
```

File: md/customize-living-application-theme.md [hunk 17]
```diff
@@ -0,0 +1,327 @@
...
+						</configuration>
+					</execution>
+				</executions>
+			</plugin>
+		</plugins>
+	</build>
+</project>
+```
+
+The `artifactId`, `name` and `description` are used to defined the theme metadata (the name used in the URL, the display name, and a description) in the `page.properties`. NodeJS and NPM version are fixed in the `properties` section. The `frontend-maven-plugin` will locally install and use these versions even if you have  NodeJS and NPM already installed in your environment. See the [plugin github repository](https://github.com/eirslett/frontend-maven-plugin) for more informations.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
The `artifactId`, `name` and `description` are used to define the theme metadata (the name used in the URL, the display name, and a description) in the `page.properties`. NodeJS and npm versions are fixed in the `properties` section. The `frontend-maven-plugin` will locally install and use these versions even if you have NodeJS and npm already installed in your environment. See the [plugin github repository](https://github.com/eirslett/frontend-maven-plugin) for more information.
```

File: md/customize-living-application-theme.md [hunk 18]
```diff
@@ -0,0 +1,327 @@
...
+//Bonita variables
+@import "bonita_variables";
+
+// Bootstrap
+@import "bootstrap";
+
+@import "bonita_pager";
+```
+The main.scss is the aggregation of 3 imports:
+* `@import "bonita_variables";` import the content of `src/scss/_bonita_variables.scss` file.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
* `@import "bonita_variables";` imports the content of `src/scss/_bonita_variables.scss` file.
```

File: md/customize-living-application-theme.md [hunk 19]
```diff
@@ -0,0 +1,327 @@
...
+* `@import "bootstrap";` import the bootstrap-sass stylesheet. You may look its content in `node_modules/bootstratp-sass/assets/stylesheets/_bootstrap.scss`.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
* `@import "bootstrap";` imports the bootstrap-sass stylesheet. You may look its content in `node_modules/bootstratp-sass/assets/stylesheets/_bootstrap.scss`.
```

File: md/customize-living-application-theme.md [hunk 20]
```diff
@@ -0,0 +1,327 @@
...
+* `@import "bonita_pager";` import the content of `src/scss/_bonita_pager.scss` file, a custom style for bootstrap pager used by the Bonita theme.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
* `@import "bonita_pager";` imports the content of `src/scss/_bonita_pager.scss` file, a custom style for Bootstrap pager used by the Bonita theme. 
```

File: md/customize-living-application-theme.md [hunk 21]
```diff
@@ -0,0 +1,327 @@
...
+  "description": "My custom theme description",
+  "license": "GPL-2.0-or-later",
+  "scripts": {
+    "build": "node-sass --precision 8 --output-style compressed --omit-source-map-url true --include-path ./node_modules/bootstrap-sass/assets/stylesheets/ src/scss/main.scss target/theme.noprefix.css && postcss target/theme.noprefix.css --no-map --use autoprefixer -b \"last 2 versions\" -o dist/theme.css"
+  },
+  "devDependencies": {
+    "node-sass": "4.11.0",
+    "postcss-cli": "6.1.2",
+    "autoprefixer": "9.5.0",
+    "bootstrap-sass": "3.3.7" //Supported version of bootstrap in Bonita
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
    "bootstrap-sass": "3.3.7" //Supported version of Bootstrap in Bonita
```

File: md/customize-living-application-theme.md [hunk 22]
```diff
@@ -0,0 +1,327 @@
...
+@import "bootstrap";
+
+@import "bonita_pager";
+```
+The main.scss is the aggregation of 3 imports:
+* `@import "bonita_variables";` import the content of `src/scss/_bonita_variables.scss` file.
+* `@import "bootstrap";` import the bootstrap-sass stylesheet. You may look its content in `node_modules/bootstratp-sass/assets/stylesheets/_bootstrap.scss`.
+* `@import "bonita_pager";` import the content of `src/scss/_bonita_pager.scss` file, a custom style for bootstrap pager used by the Bonita theme. 
+
+When using Sass you can split your stylesheets into _partials_. This is a great way to modularize your CSS and help keep things easier to maintain. A partial is simply a Sass file named with a leading underscore. You might name it something like `_partial.scss`. The underscore lets Sass know that the file is only a partial file and that it should not be generated into a CSS file. Sass partials are used with the `@import` directive like in our `src/scss/main.scss`.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
When using Sass, you can split your stylesheets into _partials_. This is a great way to modularize your CSS and help keep things easier to maintain. A partial is simply a Sass file named with a leading underscore. You might name it something like `_partial.scss`. The underscore lets Sass know that the file is only a partial file and that it should not be generated into a CSS file. Sass partials are used with the `@import` directive like in our `src/scss/main.scss`. 
```

File: md/customize-living-application-theme.md [hunk 23]
```diff
@@ -0,0 +1,327 @@
...
+
+$panel-success-heading-bg: $brand-success;
+$panel-info-heading-bg: $brand-info;
+$panel-warning-heading-bg: $brand-warning;
+$panel-danger-heading-bg: $brand-danger;
+
+//Glyphicons fonts
+$icon-font-path: "./fonts/"; // path relative to the theme.css file in the dist folder
+```
+All the variables defined in this file are used by *bootstrap-sass*. You can look the following scss file `node_modules/bootstratp-sass/assets/stylesheets/bootstrap/_variable.scss` to discover all available variables.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
All the variables defined in this file are used by *bootstrap-sass*. You can look at the following SCSS file `node_modules/bootstratp-sass/assets/stylesheets/bootstrap/_variable.scss` to discover all available variables.
```

File: md/customize-living-application-theme.md [hunk 24]
```diff
@@ -0,0 +1,327 @@
...
+
+::: info
+**Note:** Only variables declared with the `!default` flag can be overridden.
+:::
+
+SASS and SCSS have a lots of other interesting features that you should use. Check [SASS documentation](https://sass-lang.com/documentation) to known more.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
Sass and SCSS have lots of other interesting features that you could use. Check [Sass documentation](https://sass-lang.com/documentation) to known more.
```

File: md/customize-living-application-theme.md [hunk 25]
```diff
@@ -0,0 +1,327 @@
...
+                                                                          
+## Building, Deploying, Previewing a Theme
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
## Building, Deploying, and Previewing a theme
```

File: md/customize-living-application-theme.md [hunk 26]
```diff
@@ -0,0 +1,327 @@
...
+
+During the development phase you can have preview your theme using the provided test page in `test/index.html`.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
During the development phase, you can preview your theme using the provided test page in `test/index.html`.
```

File: md/customize-living-application-theme.md [hunk 27]
```diff
@@ -0,0 +1,327 @@
...
+First you will have to `build` your theme: Right click on your theme project > Build or press `Ctrl+Shift+B` shortcut.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
First you will have to `build` your theme: right click on your theme project > Build or press `Ctrl+Shift+B` for a shortcut.
```

File: md/customize-living-application-theme.md [hunk 28]
```diff
@@ -0,0 +1,327 @@
...
+
+If your theme is already associated to a living application you can just `deploy` your theme:  Right click on your theme project > Deploy or press `Ctrl+Shift+D` shortcut.
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
If your theme is already associated to a living application, you can just `deploy` your theme:  right click on your theme project > Deploy or press `Ctrl+Shift+D` for a shortcut.
```

File: md/customize-living-application-theme.md [hunk 29]
```diff
@@ -0,0 +1,327 @@
...
+
+## Create a Theme project from an existing theme in production
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
## Create a theme project from an existing theme in production
```

File: md/customize-living-application-theme.md [hunk 30]
```diff
@@ -0,0 +1,327 @@
...
+
+If you are upgrading from a previous Bonita version you may already have a theme.css file packaged in a custom page archive.  
+The easiest way of integrating your theme as a new Theme project is:
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
The easiest way of integrating your theme as a new theme project is:
```

File: md/customize-living-application-theme.md [hunk 31]
```diff
@@ -0,0 +1,327 @@
...
+* Create a New Theme
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
* 1. Create a new theme
```

File: md/customize-living-application-theme.md [hunk 32]
```diff
@@ -0,0 +1,327 @@
...
+* Extract the theme.css file from your Theme custom page zip
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
* 2. Extract the theme.css file from your theme custom page .zip archive
```

File: md/customize-living-application-theme.md [hunk 33]
```diff
@@ -0,0 +1,327 @@
...
+* Replace the content of the main.scss file with the content of the extracted theme.css file
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
* 3. Replace the content of the main.scss file with the content of the extracted theme.css file
```

File: md/customize-living-application-theme.md [hunk 34]
```diff
@@ -0,0 +1,327 @@
...
+* Retrieve all the related assets if any and copy them in the `dist` folder accordingly
```
Comment (NathalieC) [inline md/customize-living-application-theme.md]:
```suggestion
* 4. Retrieve all the related assets if any and copy them in the `dist` folder accordingly
```

**Pull request merged**